### DISCUSSION TITLE
auto import Link Image not working vscode

### DISCUSSION THREAD

**Discussion by SamyZog** [vercel/next.js#30952] [Help] [answered]
I set up a nextjs project with typescript, and vscode will not auto import `Link` or `Image` components, I have to add them manually.

Also if I import `Link` in one module I get autocompletion for the `Link` component in another module but the path shown is wrong `next/dist/client/link` it should be `next/link`

Is a nextjs issue? or a vscode issue?

**Comment by brianlovin** [comment]
I hit this all the time 😅 Typeahead importing works for all my other components and npm packages, except these components from Next.

**Comment by Joshua-Chu** [comment]
@SamyZog do you happen to find any solution for this problem? How about you @brianlovin ?

**Reply by SamyZog** [reply]
unfortunately no

**Comment by franco-roura** [comment, marked as answer]
Add a typescript declaration file with the following content
```ts
// someFile.d.ts
import Image from 'next/image'

declare module 'next/image' {
  export default Image = Image
}
```

Make sure it's matched by the "include" array in your tsconfig.json file.

You should start seeing this 
![image](https://user-images.githubusercontent.com/58178352/145413434-c2fe7e7e-18a5-4a15-a50d-6c49ee48222b.png)

You can do that with `next/link` too.

**Reply by tofuwu-dev** [reply]
Hi, Sorry but it doesn't work for me...

**Reply by WhyouDev** [reply]
![Screenshot 2024-01-04 115338](https://github.com/vercel/next.js/assets/8694548/ed7161cf-4015-4fb7-9f66-6e72bb32a5a9)
![image](https://github.com/vercel/next.js/assets/8694548/c78e17ba-f0be-4519-afc0-3959abcb411e)
It's working! just add on these sections "export default Image" without a quote to "export default Image = Image". Like what I did on the screenshot. These files on C:bla..bla..bla\node_modules\next\image.d.ts

**Reply by sidvud98** [reply]
@WhyouDev Still not working

**Comment by baranelitez** [comment]
Hi. Any idea why this happens, why it won't work, and why we have to do this type declaration? I'm using Webstorm but having the same issue.

Asking to see if there's some wrong configuration on my side or this is a thing everybody deals with? This discussion having only 3 answers made me think it might be something I'm doing wrong.

**Reply by franco-roura** [reply]
Usually, this happens when you have an unnamed default export. Compare the following:

Unnamed
<img width="852" alt="image" src="https://user-images.githubusercontent.com/58178352/155206882-b092ae2f-d0fe-43c4-b6d8-769ec080eb99.png">

Named
<img width="830" alt="image" src="https://user-images.githubusercontent.com/58178352/155206962-468c9c53-b7f6-45d1-bb6b-ec42c137175a.png">

It's just the Typescript server on your IDE not being able to find where you want to get that variable from until you name the default export. This can happen either in the source code or using module augmentation with an additional .d.ts file, in the end, it's about helping your IDE find the variable.

In the case of next/image, they're doing `export default function Image() {}`  https://github.com/vercel/next.js/blob/75b7a57e0f0044d9315eb6adbd4231b67799d0b1/packages/next/client/image.tsx#L298

I'm not sure why most IDEs can't find it, but until they change/fix it, adding a declaration file works fine.

**Comment by lelabo-m** [comment]
I have the same problem in VScode, tried the suggested solution but VScode thinks `<Image>` is of type `HTMLImageElement`, thus not suggesting other imports (even with a `.d.ts` file).

Any suggestions ?

**Reply by franco-roura** [reply]
@lelabo-m 
Does your `tsconfig.json` file have your newly created `d.ts` file in this array? Remember to restart VSCode after updating `tsconfig.json`.
<img width="356" alt="Untitled" src="https://user-images.githubusercontent.com/58178352/174322968-d0e51ba5-00b3-4de6-be0f-1327aa9eea2a.png">

**Reply by lelabo-m** [reply]
I double checked with an explicit path to the `file.d.ts`.

But I still get the following:
![Screenshot 2022-06-17 190232](https://user-images.githubusercontent.com/7324915/174346525-a2c5832a-4e11-41cb-a3b0-be63c142beb9.jpg)
![Screenshot 2022-06-17 190640](https://user-images.githubusercontent.com/7324915/174346533-b7e165d8-4873-487c-98c9-6c3b509a7ead.jpg)

**Reply by franco-roura** [reply]
Sorry, I noticed what you mean now, with my solution you can't do `Ctrl+.` or `Cmd+.` after you've finished writing the `<Image />` tag, it just suggests auto-importing it from `next/image` while you're typing it, before closing the JSX tag.

<img width="295" alt="image" src="https://user-images.githubusercontent.com/58178352/174596844-e112fae1-5433-447f-962a-184a5032d983.png">
<img width="573" alt="image" src="https://user-images.githubusercontent.com/58178352/174597488-392e76ae-99d8-421f-9341-ce0410c67ed6.png">

**Comment by Awsome-obedeh** [comment]
I am having same issue, I need help
